design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+
+### issuing
+
+This controller will copy the `status.certificate` and `status.ca` fields from
+valid CertificateRequest resources and the `tls.key` from the
+`status.nextPrivateKeySecretName` Secret resource into the `spec.secretName`.
+
+It is responsible for encoding the private key and certificate data into the
+appropriate format.
+
+Once the key-pair has been written to `spec.secretName`, it will set the
+`status.revision` field to that of the CertificateRequest and remove the
+`Issuing` status condition (in the same Update call).
+
+It will behave as follows:
+
+* If the 'Issuing' condition is `True`:
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Do nothing
+    * Find the CertificateRequest for `status.revision` + 1
+        * If none exist, do nothing.
+        * If multiple exist, do nothing (`requestmanager` will handle this)
+    * If the CertificateRequest is not in a 'final' state, do nothing.
+    * If it is 'valid':
+        * Verify the public key of the certificate request matches the public key
+          of the stored `status.nextPrivateKeySecretName`
+            * If not, do nothing (`requestmanager` will handle this)
+        * Verify the CSR options match what is requested in `certificate.spec`
+            * If not, do nothing (`requestmanager` will handle this)
+        * Encode and issue the key-pair (store it in the Secret)
+        * Set `status.revision` to revision of the CertificateRequest
+        * Remove `Issuing` status condition
+        * Clear `status.lastFailureTime` (if set)
+        * Log an Event
+    * If not 'valid':
+        * Set `status.lastFailureTime` (if not equal to CertificateRequest)
+        * Set `Issuing` status condition to `False` with reason explaining when
+          the request will be retried
+* If the 'Issuing' condition is `False` or not set:
+    * Do nothing
+
+#### trigger
+
+The trigger plugin is responsible for observing the state of the currently
+issued `spec.secretName` and the rest of the `certificate.spec` fields to
+determine whether a re-issuance is required.
+It triggers re-issuance by adding the `Issuing` status condition when a new
+certificate is required.
+
+These conditions will cause a request to be triggered:
+
+* `spec.secretName` does not exist, or contains data that cannot be decoded.
+* Issuer name/kind annotations on Secret resource do not match current `spec.issuerRef`.
+* The public key of the `tls.key` and `tls.crt` do not match.
+* A CertificateRequest for the current `status.revision` exists and the
+  current requested `certificaate.spec` does not match the options on the CSR.
+* If the stored x509 certificate's `NotAfter` is within `spec.renewBefore` of
+  the current time.
+
+Additionally, the controller is responsible for implement a 'back-off' if
+CertificateRequest resources persistently fail to complete.
+For now, this will be a continuation of the current behaviour of backing off
+by 1 hour after a request fails.
```

Comment:
Do we want to make this configurable?

Reply:
This is the current behaviour and is not a change from what we do today. Yes, this would be good to have configurable, although I actually think we're better to invest time in allowing this backoff to be *exponential* instead of a fixed period of time.